spec/models/user_badge_spec.rb

```diff
@@ -14,6 +14,16 @@
     it { is_expected.to validate_uniqueness_of(:badge_id).scoped_to(:user_id) }
   end
 
+  describe '#save' do
+    it "triggers the 'user_badge_granted' DiscourseEvent" do
+      event = DiscourseEvent.track(:user_badge_granted) do
```

Comment:
We could probably put this in a `let`.

By the way wouldn’t it be simpler to just expect the proper message with the proper args is sent here instead of having to call the underlying code?

Reply:
What benefit does putting the code here in a let bring? I prefer to keep this as it is for readability.

Will you be able to provide an example of expecting the proper message here? I’m not quite sure I understand the suggestion here to simplify the code.

Reply:
> On just expecting the proper message, this is how I’d do it normally:

I don’t quite agree with using mocks here and in general because there is no good reason for us to. In terms of testing, what we want is to treat the method as a black box. The test should not know anything about the internal implementation of the method and instead only assert for the expected outcome. Testing for internal implementation details make the tests brittle to changes in the long run.

Reply:
> DiscourseEvent is a class with a known public API, it should be properly tested too so we already know what happens when we call #trigger on it. We can then simply verify we’re sending the proper message with the proper args.

Why I say we’re testing for internal implementation detail here by mocking is because we have to assume that #trigger is the API being used. This assumption makes the test brittle in the future. For example, DiscourseEvent might add another method that allows us to trigger an event. #trigger_async for example. The expected outcome stays the same because my desired event is triggered irregardless of the way it was triggered. By mocking here, we’re unnecessarily tying ourselves to a single API making the test brittle in the long run.

Reply:
While I don’t think it will change your mind on the matter, the best thing I can do I think is to quote some excerpt from POODR (Chapter 9, Designing Cost-Effective Tests) as the author explains things way better than me :sweat_smile: 

> _The general rule is that objects should make assertions about state only for messages in their own public interfaces. Following this rule confines tests of message return values to a single place and removes unnecessary duplication, DRYing out your tests and lowering maintenance costs._
> 
> _The fact that you need not test outgoing messages for state does not mean outgoing messages need no tests at all. There are two flavors of outgoing messages, and one of them requires a different kind of test._
> 
> _Some outgoing messages have no side effects and thus matter only to their senders. The sender surely cares about the result it gets back (why else send the message?), but no other part of the application cares if the message gets sent. Outgoing messages like this are known as queries and they need not be tested by the sending object. Query messages are part of the public interface of their receiver, which already implements every necessary test of state._
> 
> _However, many outgoing messages do have side effects (a file gets written, a database record is saved, an action is taken by an observer) upon which your application depends. These messages are commands and it is the responsibility of the sending object to prove that they are properly sent. Proving that a message gets sent is a test of behavior, not state, and involves assertions about the number of times, and with what arguments, the message is sent._
> 
> _Here, then, are the guidelines for what to test: Incoming messages should be tested for the state they return. Outgoing command messages should be tested to ensure they get sent. Outgoing query messages should not be tested._
> 
> _As long as your application’s objects deal with one another strictly via public interfaces, your tests need know nothing more. When you test this minimal set of messages, no change in the private behavior of any object can affect any test. When you test outgoing command messages only to prove they get sent, your loosely coupled tests can tolerate application changes without being forced to change in turn. As long as the public interfaces remain stable, you can write tests once and they will keep you safe forever._

Reply:
Yes yes I know, “when the side effect is cheap and stable, break the rule” :slightly_smiling_face: Maybe here it’s not the best example to use mocks over a side-effect, I’m probably more used to use them and so I find this solution cheaper :wink: